CoseIndirectSignature/CoseHashV.cs

```diff
@@ -0,0 +1,330 @@
+        byte[] hash = stream != null ? hashAlgorithm.ComputeHash(stream) : hashAlgorithm.ComputeHash(data);
+
+        // handle the case where the algorithm we derived did not match the algorithm that was used to populate the CoseHashV instance.
+        return hash.Length != HashValue.Length
+            ? throw new CoseSign1Exception($@"The computed hash length of {hash.Length} for hash type {hashAlgorithm.GetType().FullName} created a hash different than the length of {HashValue.Length} which is unexpected.")
+            : hash.SequenceEqual(HashValue);
+    }
+
+    /// <summary>
+    /// Writes the current CoseHashV instance to a cbor byte[].
+    /// </summary>
+    /// <returns>a byte[] cbor representation of the CoseHashV object.</returns>
+    public byte[] Serialize()
+    {
+        CborWriter writer = new(CborConformanceMode.Strict, allowMultipleRootLevelValues: true);
+
+        writer.Reset();
+        writer.WriteInt64((long)Algorithm);
+        writer.WriteByteString(HashValue);
+        if (Location != null)
+        {
+            writer.WriteTextString(Location);
+        }
+        if (Any != null)
+        {
+            writer.WriteByteString(Any);
+        }
+
+        return writer.Encode();
+    }
+
+    /// <summary>
+    /// Reads a COSE_Hash_V structure from the <see cref="CborReader"/>.
+    /// </summary>
+    /// <param name="data">A byte[] which represents a CoseHashV object.</param>
+    /// <returns>A proper COSE_Hash_V structure if read from the reader.</returns>
+    /// <exception cref="CoseSign1Exception">Thrown if an invalid object state or format is detected.</exception>
+    public static CoseHashV Deserialize(byte[] data) => Deserialize(new CborReader(data, allowMultipleRootLevelValues: true));
+
+    /// <summary>
+    /// Reads a COSE_Hash_V structure from the <see cref="CborReader"/>.
+    /// </summary>
+    /// <param name="reader">The CBOR reader to be read from, it must have allowMultipleRootLevelValues set to true.</param>
+    /// <returns>A proper COSE_Hash_V structure if read from the reader.</returns>
+    /// <exception cref="ArgumentNullException">Thrown if <paramref name="reader"/> is null.</exception>
+    /// <exception cref="CoseSign1Exception">Thrown if an invalid object state or format is detected.</exception>
+    public static CoseHashV Deserialize(CborReader reader)
+    {
+        if (reader == null)
+        {
+            throw new ArgumentNullException(nameof(reader));
+        }
+        CoseHashV returnValue = new CoseHashV();
+
+        // CBor encodes positive or negative, so we need to check for both.
+        if (reader.PeekState() != CborReaderState.UnsignedInteger &&
+            reader.PeekState() != CborReaderState.NegativeInteger)
+        {
+            throw new CoseSign1Exception($"Invalid COSE_Hash_V structure, peek state {reader.PeekState()} was not {nameof(CborReaderState.NegativeInteger)} or  {nameof(CborReaderState.UnsignedInteger)}");
```

Comment:
RFC 9054 defines the algorithm identifier type as "int / tstr". Is it worth trying to support that construction, even if the tool can't compute a hash with an unknown algorithm? I guess in other words, should there be a way to pass a raw hash value with an arbitrary algorithm, ensure that the encoding is well-formed, but leave hash matching to an application/future validator?

Reply:
Yes, I saw it proposed as an "int / tstr".  The int mapping seemed quite a bit easier to map to existing COSE Algorithms, but we could also try to support tstr mapping of the same based on its type.  I'll push an update which attempts to support both.

Reply:
latest update addreses both int/tstr handling.

Reply:
I've added a `disableValidation` flag to both the constructor and the deserialize methods to allow objects to be fetched or created in this state.